### PR TITLE
Task1 selenium wd

### PR DESCRIPTION
**Pull request by Zhukova-Iryna**

### CODE REVIEW

File: src/main/java/BasePage.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.PageFactory;
+
+public class BasePage {
```
Comment (DmitriyBondarenko) [inline src/main/java/BasePage.java]: Please, make it `abstract`

File: src/main/java/BasePage.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    protected WebDriver driver;
+
+    public BasePage(WebDriver driver) {
+        this.driver = driver;
+        PageFactory.initElements(driver, this);
+    }
+
+    // Enter text to text field
```
Comment (DmitriyBondarenko) [inline src/main/java/BasePage.java]: In a remote repository there are should not be any comments. Please, remove all comments in your code.
You should have propper variable/method/class name for clear understanding why it is needed.

File: src/main/java/BasePage.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    // Get text from web element
+    public String getTextFromElement(WebElement webElement) {
+        return webElement.getText();
+    }
+
+
+    // Click on web element
+    public static void clickWebElement(WebElement webElement){
```
Comment (DmitriyBondarenko) [inline src/main/java/BasePage.java]: Please, keep the code style. Always use `Ctrl+Alt+L` for code formatting.

File: src/main/java/MainPage.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+public class MainPage extends BasePage {
+
+    @FindBy(xpath = ".//div[@id = 'auth-user-links']")
+    private WebElement loginLink;
+
+    @FindBy(id = "header-user-link")
```
Comment (DmitriyBondarenko) [inline src/main/java/MainPage.java]: You should use only one pattern for elelment locators. So please, use only **XPath** for element locators

File: src/main/java/MainPage.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    @FindBy(xpath = "//div[@class='right-block-inputs']//div[2]//div[1]")
+    WebElement blankRegEmailFieldMessage;
+
+    @FindBy(xpath = "//div[@class='right-block-inputs']//div[3]//div[1]")
+    WebElement blankRegPasswordFieldMessage;
+
+    @FindBy(xpath = "//button[contains(@class,'ga_register_user_popup_close']")
+    WebElement closeFormButton;*/
+
+    public MainPage(WebDriver driver){
```
Comment (DmitriyBondarenko) [inline src/main/java/MainPage.java]: Code format, please

File: src/main/java/MainPage.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        super(driver);
+    }
+
+    // Fill login form and submit
+    public void fillLoginForm(String email, String password) {
+        JavascriptExecutor js = (JavascriptExecutor) driver;
```
Comment (DmitriyBondarenko) [inline src/main/java/MainPage.java]: Please, move it to some utils class and create `static` method for it.

File: src/main/java/BasePage.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+public class BasePage {
+    protected WebDriver driver;
+
+    public BasePage(WebDriver driver) {
+        this.driver = driver;
+        PageFactory.initElements(driver, this);
+    }
+
+    // Enter text to text field
+    public static void enterTextInTextField(WebElement textField, String inputText) {
```
Comment (DmitriyBondarenko) [inline src/main/java/BasePage.java]: Make it and all methods in this class `protected` if needed

File: src/main/java/SearchPage.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import java.util.List;
+
+public class SearchPage extends BasePage {
+
+    @FindBy(xpath = ".//div[@class='products fluid-grid clearer-block']")
```
Comment (DmitriyBondarenko) [inline src/main/java/SearchPage.java]: Can we use `contains()` for classes in the XPath?

File: src/main/java/SearchPage.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    WebElement searchResultBlock;
+
+    public SearchPage(WebDriver driver) {
+        super(driver);
+    }
+
+    public List<WebElement> getSearchElements() {
+        return driver.findElements(By.xpath(".//div[contains(@class, 'product-block type')]"));
```
Comment (DmitriyBondarenko) [inline src/main/java/SearchPage.java]: Can we move it to `@FindBy` formmat?

Comment (DmitriyBondarenko) [inline src/main/java/SearchPage.java]: Can we move it to `@FindBy` formmat?

File: src/test/java/BaseTest.java [hunk 10]
```diff
@@ -1,16 +1,43 @@
...
+    public void setUp(@Optional("chrome") String browser) throws Exception {
+        initDrivers(browser);
+        driver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
         driver.manage().window().maximize();
     }
+
+    private void initDrivers(String browser) {
+        switch (browser){
+            case "chrome":
+                System.setProperty("webdriver.chrome.driver", "src/main/resources/properties/chromedriver.exe");
```
Comment (DmitriyBondarenko) [inline src/test/java/BaseTest.java]: Please, move the path to drivers to Property file

File: src/test/java/BaseTest.java [hunk 11]
```diff
@@ -1,16 +1,43 @@
...
+                break;
+            case "firefox":
+                System.setProperty("webdriver.gecko.driver", "src/main/resources/properties/gecodriver.exe");
+                driver = new FirefoxDriver();
+                break;
+            case "ie":
+                System.setProperty("webdriver.ie.driver", "src/main/resources/properties/IEDriverServer.exe");
+                driver = new InternetExplorerDriver();
+                break;
+            default: throw new IllegalArgumentException("Unknown browser" + browser);
```
Comment (DmitriyBondarenko) [inline src/test/java/BaseTest.java]: `String.format()` for strings concatination

File: src/test/java/BaseTest.java [hunk 12]
```diff
@@ -1,16 +1,43 @@
...
+        }
+    }
+
+    @AfterMethod(alwaysRun = true)
```
Comment (DmitriyBondarenko) [inline src/test/java/BaseTest.java]: Do we really need `alwaysRun` parameter here?

File: src/test/java/MainPageTest.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    private String password;
+    private String regEmail;
+    private String regPassword;
+    private String empty = "";
+    private String searchRequest;
+
+    @BeforeClass
+    public void setUp() {
+        String commonProperties = "src/main/resources/test-properties.properties";
+        Properties properties = new Properties();
```
Comment (DmitriyBondarenko) [inline src/test/java/MainPageTest.java]: Please, create appropriate class utils for property reader mathod.

File: src/test/java/MainPageTest.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    @Test
+    public void checkNegativeLoginScenario() {
+        mainPage.fillLoginForm(email, regPassword);
+        Assert.assertTrue(mainPage.checkErrorLoginMessageAppears(), "Error message doesn't appears");
+    }
+
+    @Test
+    public void checkPositiveRegistrationScenario() {
+        mainPage.fillRegistrationForm(regEmail, regPassword);
+        Assert.assertTrue(driver.findElement(By.xpath(".//div[@class = 'reg-success-text']")).isDisplayed(), "Successful registration message doesn't appear");
```
Comment (DmitriyBondarenko) [inline src/test/java/MainPageTest.java]: Please, create appropriate element on the page and use it here

File: src/test/java/MainPageTest.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    @Test
+    public void checkNegativeRegistrationScenario() {
+        mainPage.fillRegistrationForm(email, regPassword);
+        Assert.assertTrue(mainPage.checkErrorLoginMessageAppears(), "Error message doesn't appears");
+    }
+
+    @Test
+    public void searchStartsTest() {
+        mainPage.runSearch(searchRequest);
+        new WebDriverWait(driver, 5).until(ExpectedConditions.presenceOfElementLocated(By.xpath(".//div[@id='page-breadcrumbs']")));
```
Comment (DmitriyBondarenko) [inline src/test/java/MainPageTest.java]: Move it to appropriate page class, please

File: src/test/resources/test-properties.properties [hunk 16]
```diff
@@ -0,0 +1,6 @@
+baseurl=https://price.ua/
```
Comment (DmitriyBondarenko) [inline src/test/resources/test-properties.properties]:
```suggestion
base_url = https://price.ua/
valid_email = ...
valid_password = ...
```

File: src/main/java/BasePage.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    public String getTextFromElement(WebElement webElement) {
+        return webElement.getText();
+    }
+
+
+    // Click on web element
+    public static void clickWebElement(WebElement webElement){
+        webElement.click();
+    }
+
```
Comment (DmitriyBondarenko) [inline src/main/java/BasePage.java]: Remove line breake please

File: src/test/java/MainPageTest.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+        baseUrl = properties.getProperty("baseurl");
+        email = properties.getProperty("validemail");
+        password = properties.getProperty("validpassword");
+        regEmail = properties.getProperty("regemail");
+        regPassword = properties.getProperty("regpassword");
+        searchRequest = properties.getProperty("searchrequest");
+        driver.get(baseUrl);
+        mainPage = new MainPage(driver);
+    }
+
```
Comment (DmitriyBondarenko) [inline src/test/java/MainPageTest.java]: Remove line breake please

**Review by DmitriyBondarenko — CHANGES_REQUESTED** [review]
Fix all comments please

File: src/main/java/BasePage.java (hunk 1 above)
Comment (Zhukova-Iryna) [reply src/main/java/BasePage.java]: Done

File: src/main/java/BasePage.java (hunk 2 above)
Comment (Zhukova-Iryna) [reply src/main/java/BasePage.java]: Done

File: src/main/java/MainPage.java (hunk 4 above)
Comment (Zhukova-Iryna) [reply src/main/java/MainPage.java]: Done

File: src/main/java/MainPage.java (hunk 5 above)
Comment (Zhukova-Iryna) [reply src/main/java/MainPage.java]: Done

File: src/test/java/BaseTest.java (hunk 12 above)
Comment (Zhukova-Iryna) [reply src/test/java/BaseTest.java]: Done

File: src/test/resources/test-properties.properties (hunk 16 above)
Comment (Zhukova-Iryna) [reply src/test/resources/test-properties.properties]: Done

File: src/main/java/ua/price/base/BaseTest.java [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+            case "firefox":
+                System.setProperty("webdriver.gecko.driver", PropertyReader.getProperty("firefox_driver"));
+                driver = new FirefoxDriver();
+                break;
+            case "ie":
+                System.setProperty("webdriver.ie.driver", PropertyReader.getProperty("ie_driver"));
+                driver = new InternetExplorerDriver();
+                break;
+            default:
+                throw new IllegalArgumentException(String.format("Unknown browser", browser));
```
Comment (DmitriyBondarenko) [inline src/main/java/ua/price/base/BaseTest.java]: `String.format("Unknown browser: %s", browser)`

File: src/main/java/ua/price/pages/MainPage.java [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+    private WebElement regPasswordField;
+
+    @FindBy(xpath = "//div[contains(@class,'mt_10')]")
+    private WebElement submitRegistration;
+
+    @FindBy(xpath = ".//div[@class = 'reg-success-text']")
+    private WebElement confirmRegistrationMessage;
+
+    @FindBy(xpath = ".//div[@id='google_ads_iframe_/6560281/Price_Promo_text_1_0__container__']")
+    WebElement popularCategoryAd;
```
Comment (DmitriyBondarenko) [inline src/main/java/ua/price/pages/MainPage.java]: `private`

File: src/main/java/ua/price/pages/MainPage.java [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+        searchFieldInput.sendKeys(searchRequest + Keys.ENTER);
+        return new SearchPage(driver);
+    }
+
+
+    public void waitUntilPageLoaded() {
+        new WebDriverWait(driver, 5).until(ExpectedConditions.visibilityOf(popularCategoryAd));
+        System.out.println("load");
+    }
+
```
Comment (DmitriyBondarenko) [inline src/main/java/ua/price/pages/MainPage.java]: Remove LB

File: src/main/java/ua/price/pages/MainPage.java [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+    public boolean checkConfirmRegistrationMessageAppears() {
+        return confirmRegistrationMessage.isDisplayed();
+    }
+
+    public SearchPage runSearch(String searchRequest) {
+        searchFieldInput.sendKeys(searchRequest + Keys.ENTER);
+        return new SearchPage(driver);
+    }
+
```
Comment (DmitriyBondarenko) [inline src/main/java/ua/price/pages/MainPage.java:93]: Remove LB

File: src/main/java/ua/price/pages/SearchPage.java [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import ua.price.base.BasePage;
+
+import java.util.List;
+
+public class SearchPage extends BasePage {
+
+    @FindBy(xpath = ".//div[@id = 'fluid-grid']")
+    WebElement searchResultBlock;
```
Comment (DmitriyBondarenko) [inline src/main/java/ua/price/pages/SearchPage.java]: All elements are `private`

File: src/main/java/ua/price/properties/PropertyReader.java [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+
+public class PropertyReader {
+    public static String getProperty(String name) {
+        try {
+            Properties props = new Properties();
+            props.load(ClassLoader.class.getResourceAsStream("/test_properties.properties"));
+
+            String property = props.getProperty(name);
+            if (property == null) {
+                throw new IllegalArgumentException("Could not read property with key " + name);
```
Comment (DmitriyBondarenko) [inline src/main/java/ua/price/properties/PropertyReader.java]: `String.format()`

File: src/main/java/ua/price/pages/MainPage.java (hunk 21 above)
Comment (Zhukova-Iryna) [reply src/main/java/ua/price/pages/MainPage.java]: Done

File: src/main/java/ua/price/pages/MainPage.java (hunk 20 above)
Comment (Zhukova-Iryna) [reply src/main/java/ua/price/pages/MainPage.java]: Done

File: src/main/java/ua/price/pages/MainPage.java (hunk 22 above)
Comment (Zhukova-Iryna) [reply src/main/java/ua/price/pages/MainPage.java:93]: Done

File: src/main/java/ua/price/pages/SearchPage.java (hunk 23 above)
Comment (Zhukova-Iryna) [reply src/main/java/ua/price/pages/SearchPage.java]: Done

File: src/main/java/ua/price/properties/PropertyReader.java (hunk 24 above)
Comment (Zhukova-Iryna) [reply src/main/java/ua/price/properties/PropertyReader.java]: Done

File: src/main/java/ua/price/base/BaseTest.java (hunk 19 above)
Comment (Zhukova-Iryna) [reply src/main/java/ua/price/base/BaseTest.java]: Done

File: src/test/java/MainPageTest.java (hunk 15 above)
Comment (Zhukova-Iryna) [reply src/test/java/MainPageTest.java]: Done

File: src/test/java/MainPageTest.java (hunk 14 above)
Comment (Zhukova-Iryna) [reply src/test/java/MainPageTest.java]: Done

File: src/test/java/MainPageTest.java (hunk 13 above)
Comment (Zhukova-Iryna) [reply src/test/java/MainPageTest.java]: Done

File: src/test/java/BaseTest.java (hunk 11 above)
Comment (Zhukova-Iryna) [reply src/test/java/BaseTest.java]: Done

File: src/test/java/BaseTest.java (hunk 10 above)
Comment (Zhukova-Iryna) [reply src/test/java/BaseTest.java]: Done

File: src/main/java/SearchPage.java (hunk 9 above)
Comment (Zhukova-Iryna) [reply src/main/java/SearchPage.java]: Done

File: src/main/java/SearchPage.java (hunk 8 above)
Comment (Zhukova-Iryna) [reply src/main/java/SearchPage.java]: Done

File: src/main/java/BasePage.java (hunk 7 above)
Comment (Zhukova-Iryna) [reply src/main/java/BasePage.java]: Done

**Review by DmitriyBondarenko — APPROVED** [review]

**Pull request merged**